compiler/rustc_middle/src/query/mod.rs

```diff
@@ -177,6 +177,12 @@ rustc_queries! {
         separate_provide_extern
     }
 
+    query opaque_ty_origin(key: DefId) -> hir::OpaqueTyOrigin
+    {
+        desc { "get the type of opaque (RPIT, async, or TAIT)" }
+        separate_provide_extern
+    }
```

Comment:
This is a footgun, there's now `TyCtxt::opaque_ty_origin` and `InferCtxt::opaque_ty_origin` where the latter only returns `Some` if the defining scope rules were checked and the opaque type is in its defining scope. I worry we'll accidentally end up not checking that somewhere. Looking at the cross-crate use sites of this, maybe just create a `query is_type_alias_impl_trait(key: DefId) -> bool` and keep everything else using `LocalDefId`s